### DISCUSSION TITLE
Set AWS credentials for aws-machine-image (AMI) datasource

### DISCUSSION THREAD

**Discussion by BrunnerLivio** [renovatebot/renovate#13364] [Request Help] [answered]
I am currently experimenting with the aws-machine-image datasource, though I always get the error that the region is not set.

I am running Renovate in a self-hosted Gitlab CI Pipeline using the latest Docker image (`sha256:a71762e5f6e309063e54b7c5460d029b3cdc355e45837a8d4db7f65fe230fa63 for renovate/renovate:latest`) and I also set the AWS Credentials via environment variables (`AWS_DEFAULT_REGION`, `AWS_ACCESS_ID`, `AWS_SECRET_KEY`) in the self-hosted pipeline as well as the target repository. Am I missing something? Where shall the credentials be allocated?

In the Pipeline logs where Renovate gets executed:

```shell
# NOTE: env variable is available
$ echo $AWS_DEFAULT_REGION
us-west-2
$ renovate
INFO: Repository started (repository=__REDACTED__)
       "renovateVersion": "31.15.0"
 INFO: Dependency extraction complete (repository=__REDACTED__)
       "baseBranch": "main",
       "stats": {
         "managers": {
           "gitlabci": {"fileCount": 1, "depCount": 4},
           "regex": {"fileCount": 1, "depCount": 1},
           "terraform": {"fileCount": 1, "depCount": 2}
         },
         "total": {"fileCount": 3, "depCount": 7}
       }
..
DEBUG: Datasource unknown error (repository=__REDACTED__)
       "datasource": "aws-machine-image",
       "lookupName": "[{\"Name\":\"owner-id\",\"Values\":[\"__REDACTED__\"]},{\"Name\":\"name\",\"Values\":[\"__REDACTED__\"]}]",
       "err": {
         "message": "Region is missing",
         "stack": "Error: Region is missing\n    at default (/usr/src/app/node_modules/@aws-sdk/config-resolver/src/regionConfig/config.ts:10:11)\n    at /usr/src/app/node_modules/@aws-sdk/node-config-provider/src/fromStatic.ts:9:42\n    at /usr/src/app/node_modules/@aws-sdk/property-provider/src/chain.ts:20:18\n    at coalesceProvider (/usr/src/app/node_modules/@aws-sdk/property-provider/src/memoize.ts:57:18)\n    at Object.isConstant [as region] (/usr/src/app/node_modules/@aws-sdk/property-provider/src/memoize.ts:69:20)\n    at Object.getEndpointFromRegion (/usr/src/app/node_modules/@aws-sdk/config-resolver/src/endpointsConfig/utils/getEndpointFromRegion.ts:12:18)\n    at buildHttpRpcRequest (/usr/src/app/node_modules/@aws-sdk/client-ec2/src/protocols/Aws_ec2.ts:75049:66)\n    at /usr/src/app/node_modules/@aws-sdk/middleware-serde/src/serializerMiddleware.ts:18:21\n    at /usr/src/app/node_modules/@aws-sdk/middleware-logger/src/loggerMiddleware.ts:22:22\n    at AwsMachineImageDataSource.getSortedAwsMachineImages (/usr/src/app/node_modules/renovate/lib/datasource/aws-machine-image/index.ts:52:28)\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:133:20\n    at AwsMachineImageDataSource.getReleases (/usr/src/app/node_modules/renovate/lib/datasource/aws-machine-image/index.ts:96:20)\n    at /usr/src/app/node_modules/renovate/lib/util/cache/package/decorator.ts:133:20\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:250:13)\n    at getPkgReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:302:7)\n    at lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:77:32)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:38:13)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
       }
DEBUG: Failed to look up dependency ami (repository=__REDACTED__, packageFile=infrastructure/prod/main.tf, dependency=ami)
```

**Comment by BrunnerLivio** [comment]
Maybe @markussiebert has a hunch? I looked at your [testing repo](https://github.com/markussiebert/renovate-aws-machine-image-testrepo) & I have a very similar setup, though I can't make sense where the credentials are stored :(

**Reply by markussiebert** [reply]
Hi,

As this is relying on the Javascript aws sdk, it should accept all default ways of passing credentials. I have to say, that I only tested the way via the .aws/credentials file and the eks iam roles for service accounts. Both worked flawlessly...

**Reply by markussiebert** [reply]
Can you please try specifying AWS_REGION instead of AWS_DEFAULT_REGION?

**Reply by BrunnerLivio** [reply]
Oh, that did the trick! Setting `AWS_REGION`, `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY ` as CI/CD variables on my Renovate job!

Thanks a lot for your help & your awesome PR! REALLY appreciated 🚀 🙏 

I do wonder though; would it be possible to somehow set the credentials on a repository level? Like you'd normally do with [`hostRules`](https://docs.renovatebot.com/configuration-options/#hostrules)

**Reply by markussiebert** [reply]
I never would recommend checking credentials into your git repository. But it should be possible to extend your renovate config and implement functions as part of your config that could resolve credentials for you...

**Comment by BrunnerLivio** [comment, marked as answer]
Setting `AWS_REGION`, `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` as env variables solves this problem.

E.g.

```
docker run --env AWS_REGION=us-west-2  --env AWS_ACCESS_KEY_ID=xxx --env AWS_SECRET_ACCESS_KEY=xxx renovate/renovate:latest ....
```
